src/performRequest.js

```diff
@@ -0,0 +1,161 @@
+const defaultRequest = require('request');
+const caseless = require('caseless');
+
+const defaultLogger = require('./logger');
+
+
+/**
+ * Performs the HTTP request as described in the 'transaction.request' object.
+ *
+ * In future we should introduce a 'real' request object as well so user has
+ * access to the modifications made on the way.
+ *
+ * @param {string} uri
+ * @param {Object} transactionReq
+ * @param {Object} [options]
+ * @param {Object} [options.logger] Custom logger
+ * @param {Object} [options.request] Custom 'request' library implementation
+ * @param {Object} [options.http] Custom default 'request' library options
+ * @param {Function} callback
+ */
+function performRequest(uri, transactionReq, options, callback) {
+  if (typeof options === 'function') {
+    [options, callback] = [{}, options];
+  }
+  const logger = options.logger || defaultLogger;
+  const request = options.request || defaultRequest;
+
+  const httpOptions = Object.assign({}, options.http || {});
+  httpOptions.proxy = false;
+  httpOptions.followRedirect = false;
+  httpOptions.encoding = null;
+  httpOptions.method = transactionReq.method;
+  httpOptions.uri = uri;
+
+  try {
+    httpOptions.body = getBodyAsBuffer(transactionReq.body, transactionReq.bodyEncoding);
+    httpOptions.headers = setContentLength(transactionReq.headers, httpOptions.body);
+
+    const protocol = httpOptions.uri.split(':')[0].toUpperCase();
+    logger.debug(`Performing ${protocol} request to the server under test: `
+      + `${httpOptions.method} ${httpOptions.uri}`);
+
+    request(httpOptions, (error, res, resBody) => {
+      logger.debug(`Handling ${protocol} response from the server under test`);
+      if (error) {
+        callback(error);
+      } else {
+        callback(null, createTransactionRes(res, resBody));
+      }
+    });
+  } catch (error) {
+    process.nextTick(() => callback(error));
+  }
+}
+
+
+/**
+ * Coerces the HTTP request body to a Buffer
+ *
+ * @param {string|Buffer} body
+ * @param {*} encoding
+ */
+function getBodyAsBuffer(body, encoding) {
+  return body instanceof Buffer
+    ? body
+    : Buffer.from(`${body || ''}`, normalizeBodyEncoding(encoding));
+}
+
+
+/**
+ * Returns the encoding as either 'utf-8' or 'base64'. Throws
+ * an error in case any other encoding is provided.
+ *
+ * @param {string} encoding
+ */
+function normalizeBodyEncoding(encoding) {
+  if (!encoding) {
+    return 'utf-8';
+  }
+  const lcEncoding = encoding.toLowerCase();
+  if (lcEncoding === 'utf-8' || lcEncoding === 'utf8') {
+    return 'utf-8';
+  }
+  if (lcEncoding === 'base64') {
+    return 'base64';
+  }
+  throw new Error(`Unsupported encoding: '${encoding}' (only UTF-8 and `
+    + 'Base64 are supported)');
```

Comment:
Nit: I would personally use `switch` instead `if`(s) wherever possible, something like:
```javascript
switch (lcEncoding) {
  case 'utf-8':
  case 'utf8':
    return 'utf-8'
  case 'base64';
    return 'base64';
  default:
    throw new Error(`Unsupported encoding: '${encoding}' (only UTF-8 and `
      + 'Base64 are supported)');
}
```
But it is not an issue, just an idea.

Reply:
I'll trust you this is more _javascriptonic_  😄 Python [doesn't have `switch`](https://www.quora.com/Why-does-python-not-have-a-switch-statement) so I never think of using it, even in situations where it probably makes sense.

Reply:
ae58682